python/gen_commands.py

```diff
@@ -0,0 +1,138 @@
+#!/usr/bin/env python3
+
+import dataclasses
+import pathlib
+import re
+
+
+# license header to use for generated go files
+HEADER = '''\
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//
+// Copyright (C) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+'''
+
+BASE_MODULE_NAME = 'github.com/IOTechSystems/onvif'
+
+SERVICE_NAMES = [
+    'analytics',
+    'device',
+    'event',
+    'media',
+    'media2',
+    'ptz',
+]
+
+
+@dataclasses.dataclass
+class FunctionsGenerator:
+    service_name: str
+    input_file: pathlib.Path
+    output_file: pathlib.Path
+    commands = []
+
+    @staticmethod
+    def make_function(cmd):
+        return '''\
+type %sFunction struct{}
+
+func (_ *%sFunction) Request() interface{} {
+\treturn &%s{}
+}
+func (_ *%sFunction) Response() interface{} {
+\treturn &%sResponse{}
+}
+
+''' % (cmd, cmd, cmd, cmd, cmd)
+
+    def run(self):
+        self.commands = []
+        with open(self.input_file) as f:
+            with open(self.output_file, 'w') as w:
+                w.write(HEADER)
+                w.write(f'package {self.service_name}\n\n')
+
+                while line := f.readline():
+                    if m := re.search(r'type ([^ ]+)Response struct', line):
+                        cmd = m.group(1)
+                        if cmd.endswith('Fault'):
+                            continue  # skip Fault responses
+                        self.commands.append(cmd)
+
+                self.commands = sorted(self.commands)
+                for cmd in self.commands:
+                    print(f'{self.service_name}.{cmd}')
+                    w.write(FunctionsGenerator.make_function(cmd))
+
+    def write_func_map(self, w):
+        w.write('var %sFunctionMap = map[string]Function{\n' % self.proper_name)
+        for cmd in self.commands:
+            w.write('\t%s: &%s.%sFunction{},\n' % (cmd, self.service_name, cmd))
+        w.write('}\n\n')
+
+    @property
+    def proper_name(self):
+        return (self.service_name[0].upper() + self.service_name[1:]).replace('Ptz', 'PTZ')
+
+
+@dataclasses.dataclass
+class MainGenerator:
+    root_dir: pathlib.Path
+    generators = []
+    unique_cmds = set()
+
+    def _generate_mappings(self):
+        with open(self.root_dir.joinpath('mappings.go'), 'w') as w:
+            w.write(HEADER)
+            w.write('package onvif\n\n')
+            w.write('import (\n')
+            for service_name in SERVICE_NAMES:
+                w.write(f'\t"{BASE_MODULE_NAME}/{service_name}"\n')
+            w.write(')\n\n')
+
+            for generator in self.generators:
+                generator.write_func_map(w)
+
+    def _generate_constants(self):
```

Comment:
rename to  _generate_names?

Reply:
Ah ok I see. The reason I called it constants was because it was generating the `const` values, but the file was actually named `names.go`, so yeah `_generate_names` probably makes more sense.